lib/versioncontrol/maintenancewindow/maintenancewindow.go

```diff
@@ -0,0 +1,444 @@
+	// auth connectivity. When connectivity to auth is established, a context-like value should be sent over
+	// the channel. If auth connectivity is subsequently lost, that context-like value must be canceled.
+	// During normal runtime, this should use DownstreamInventoryHandle.Sender(). During tests
+	// it can just be a stream of context.Context values.
+	AuthConnectivitySentinel <-chan C
+
+	// --- below fields are all optional
+
+	// UnhealthyThreshold is the threshold after which failure to export
+	// is treated as an unhealthy event.
+	UnhealthyThreshold time.Duration
+
+	// ExportInterval is the interval at which exports are attempted
+	ExportInterval time.Duration
+
+	// FirstExport is a custom duration used for firt export attempt.
+	FirstExport time.Duration
+
+	testEvents chan testEvent
+}
+
+func (c *ExporterConfig[C]) CheckAndSetDefaults() error {
+	if c.Driver == nil {
+		return trace.BadParameter("exporter config missing required parameter 'Driver'")
+	}
+
+	if c.ExportFunc == nil {
+		return trace.BadParameter("exporter config missing required parameter 'ExportFunc'")
+	}
+
+	if c.AuthConnectivitySentinel == nil {
+		return trace.BadParameter("exporter config missing required parameter 'AuthConnectivitySentinel'")
+	}
+
+	// allow switching to faster default duration values for testing purposes
+	fastExport := os.Getenv("TELEPORT_UNSTABLE_FAST_MW_EXPORT") == "yes"
+
+	if c.UnhealthyThreshold == 0 {
+		// 9m is fairly arbitrary, but was picked based on the the idea that a good unhealthy threshold aught to be
+		// long enough to minimize sensitivity to control plane restarts, but short enough that by the time an instance
+		// appears offline in the teleport UI, we aught to be able to assume that its unhealthy status has been propagated
+		// to its upgrader.
+		c.UnhealthyThreshold = 9 * time.Minute
+
+		if fastExport {
+			c.UnhealthyThreshold = 9 * time.Second
+		}
+	}
+
+	if c.ExportInterval == 0 {
+		// 40m is fairly arbitrary, but was picked on the basis that we want to keep exports pretty infrequent, but still frequent
+		// *enough* that one should be able to be fairly confident of full propagation to all agents within an hour, even assuming
+		// some amount of errors/retries.
+		c.ExportInterval = 40 * time.Minute
+
+		if fastExport {
+			c.ExportInterval = 40 * time.Second
+		}
+	}
+
+	if c.FirstExport == 0 {
+		// note: we add an extra millisecond since FullJitter can sometimes return 0, but our interval helpers interpret FirstDuration=0
+		// as meaning that we don't want a custom first duration. this is usually fine, but since the actual export interval is so long,
+		// it is important that a shorter first duration is always observed.
+		c.FirstExport = time.Millisecond + utils.FullJitter(c.UnhealthyThreshold/2)
+	}
+
+	return nil
+}
+
+// Exporter is a helper used to
+type Exporter[C contextLike] struct {
+	cfg          ExporterConfig[C]
+	closeContext context.Context
```

Comment:
I was pinged on a review recently and advised not to store context in structs. Seems like a reasonable idea. It'd be easy enough to adjust the call to `Run` in `service.go` to just say something like:

```
		process.RegisterCriticalFunc("maintenancewindow.export", func() error {
                    return exporter.Run(process.ExitContext())
                )
```

I don't know that `Close` would be necessary if you did this, as the context would be canceled as part of process shutdown.

Reply:
I am going to push back on this one... storing a _request scoped_ (or potentially request-scoped) context in a struct is bad, because it makes it too easy to accidentally introduce weird memory leaks.  This is because request-scoped context frequently hold onto references to lots of request-scoped date (e.g. certs).

Using `context.Background()` to create a cancellation helper for a component that does cancellable I/O is fine IMO.  Rather than bake our own alternative closure mechanism, if I were to deviate from this pattern I'd probably deviate by creating a custom `CloseContext` helper type that implements `context.Context` and never contains any context values (i.e. a context type that statically promises that it does not hold references to request-scoped values).

We could use `process.ExitContext()`, but IMO that's not a great pattern.  I've found myself leaning toward the opinion that persistent components should be explicitly rather than implicitly closed where possible.  Mostly because this allows control of the _order_ of closure.  Context cancellation cancels all layers simultaneously, which has caused confusing errors for us in the past.  Mostly a moot point here, since no other components rely on exporter being healthy, but it's a pattern I'm trying to force myself to stick to.